src/vmm/src/arch/x86_64/mod.rs

```diff
@@ -128,17 +142,81 @@ pub fn initrd_load_addr(guest_mem: &GuestMemoryMmap, initrd_size: usize) -> Opti
     Some(align_to_pagesize(lowmem_size - initrd_size) as u64)
 }
 
+/// Configures the system for booting Linux.
+pub fn configure_system_for_boot(
+    vmm: &mut Vmm,
+    vcpus: &mut [Vcpu],
+    machine_config: &MachineConfig,
+    cpu_template: &CustomCpuTemplate,
+    entry_point: EntryPoint,
+    initrd: &Option<InitrdConfig>,
+    boot_cmdline: Cmdline,
+) -> Result<(), ConfigurationError> {
+    // Construct the base CpuConfiguration to apply CPU template onto.
+    let cpu_config = {
+        let cpuid = cpuid::Cpuid::try_from(vmm.kvm.supported_cpuid.clone())
+            .map_err(GuestConfigError::CpuidFromKvmCpuid)?;
+        let msrs = vcpus[0]
+            .kvm_vcpu
+            .get_msrs(cpu_template.msr_index_iter())
+            .map_err(GuestConfigError::VcpuIoctl)?;
+        CpuConfiguration { cpuid, msrs }
+    };
+
+    // Apply CPU template to the base CpuConfiguration.
+    let cpu_config = CpuConfiguration::apply_template(cpu_config, cpu_template)?;
+
+    let vcpu_config = VcpuConfig {
+        vcpu_count: machine_config.vcpu_count,
+        smt: machine_config.smt,
+        cpu_config,
+    };
+
+    // Configure vCPUs with normalizing and setting the generated CPU configuration.
+    for vcpu in vcpus.iter_mut() {
+        vcpu.kvm_vcpu
+            .configure(vmm.guest_memory(), entry_point, &vcpu_config)
+            .map_err(ConfigurationError::VcpuConfigure)?;
+    }
+
+    // Write the kernel command line to guest memory. This is x86_64 specific, since on
+    // aarch64 the command line will be specified through the FDT.
+    let cmdline_size = boot_cmdline
+        .as_cstring()
+        .map(|cmdline_cstring| cmdline_cstring.as_bytes_with_nul().len())
+        .expect("Cannot create cstring from cmdline string");
+
+    load_cmdline::<GuestMemoryMmap>(
+        vmm.guest_memory(),
+        GuestAddress(crate::arch::x86_64::layout::CMDLINE_START),
+        &boot_cmdline,
+    )
+    .map_err(ConfigurationError::LoadCommandline)?;
+    configure_system(
+        &vmm.guest_memory,
+        &mut vmm.resource_allocator,
+        GuestAddress(crate::arch::x86_64::layout::CMDLINE_START),
+        cmdline_size,
+        initrd,
+        vcpu_config.vcpu_count,
+        entry_point.protocol,
+    )?;
+
+    // Create ACPI tables and write them in guest memory
+    // For the time being we only support ACPI in x86_64
+    create_acpi_tables(
+        &vmm.guest_memory,
+        &mut vmm.resource_allocator,
+        &vmm.mmio_device_manager,
+        &vmm.acpi_device_manager,
+        vcpus,
+    )
+    .map_err(ConfigurationError::Acpi)?;
```

Comment:
this one already has a `#[from]`, so we can just drop the `.map_err()`

Reply:
fixed